Add new blog post: How LY Corporation Utilizes MLflow (#408)
mlflow/mlflow-website: website/blog/2025-11-25-ly-croporation-utlizes-mlflow/index.md

---
title: "Introducing How LY Corporation Utilizes MLflow"
slug: ly-corporation-utilizes-mlflow
tags: [mlflow, authorization]
authors: [motoki-yuhara,  mlflow-maintainers]
thumbnail: img/blog/ly-utilizes-mlflow/AIPF.png
---

# How LY Corporation Uses MLflow: An Overview
Hello, I’m Motoki Yuhara from [LY Corporation](https://www.lycorp.co.jp/en/company/overview/), and I’m responsible for building and operating the Managed MLflow service for our internal AI platform.

> LY Corporation is one of Japan’s major tech companies, running various online services such as advertising, e-commerce, and digital membership platforms. The company also manages several group companies, driving innovation across different digital domains.

Reviewer: ```suggestion
LY Corporation is one of Japan’s leading tech companies, operating a wide range of online services including advertising, e-commerce, and digital membership platforms. Through its group companies, LY continues to advance innovation across different digital domains.
```

At LY Corporation, we have developed an in-house AI platform that supports the entire lifecycle of AI and machine learning–based services (MLOps) across more than 100 products and services.

This platform provides end-to-end support for the MLOps lifecycle, from model development to deployment and operation.

> 

Reviewer: ```suggestion
```

@yukimori Can we reduce the number of line breaks a bit to avoid using extra spaces?

As part of this platform, we use MLflow as a core component for managing model training, evaluation, and model lifecycle management.

> 

Reviewer: ```suggestion
This platform provides end-to-end MLOPs capabilities from model development to deployment and operation. MLflow is a core component of the platform for managing model training, evaluation, and model lifecycle.

```

![Aipf](/img/blog/ly-utilizes-mlflow/AIPF.png)

> # Managed MLflow Service on Kubernetes

Reviewer: ```suggestion
## Managed MLflow Service on Kubernetes
```

The level-1 section is reserved for the blog title. Can we update the other sections to have level-2?

When each user independently builds their own MLflow server, it can lead to inefficient use of computing resources.
To improve efficiency, we decided to provide MLflow as a managed service running on our Kubernetes-based computing infrastructure.

Since multiple services within LY Corporation use MLflow, we needed to ensure both stability and data isolation.
To achieve this, we provide a dedicated MLflow server instance for each service.

> However, this also means that we must carefully control access so that only the project members of a given service can use its corresponding MLflow instance.

Reviewer: ```suggestion
Since multiple services within LY Corporation use MLflow, we needed to ensure both stability and data isolation.
To achieve this, we provide a dedicated MLflow server instance for each service.
However, this also requires strict access control so that only the authorized project members of a given service can use the corresponding MLflow instance.
```

# Implementing Service-to-Service Authentication and Authorization Based on OAuth 2.0
In OAuth 2.0, the standard specification for authorization, the Client Credentials Grant Flow is defined as the authentication and authorization mechanism for service-to-service communication.

This flow allows a service to authenticate itself—without any end-user involvement—obtain an access token, and use that token to call protected APIs.

The flow works as follows:

```
[Service A] → Authorization Server (authenticates using client_id and client_secret)
        ↓
[Authorization Server] issues an Access Token
        ↓
[Service A] → [Service B (Resource Server)] accesses API (Access Token validation and authorization check)
```

Applying this flow to our environment, the relationship between the model training program and MLflow can be described as follows:

The model training program (corresponding to Service A) obtains an access token and attaches it to API requests via the Authorization header.

> MLflow (corresponding to Service B / Resource Server) validates the access token included in the request and performs authorization checks based on configured policies.

Reviewer: ```suggestion
1. The model training program (corresponding to Service A) obtains an access token and attaches it to API requests via the Authorization header.
2. MLflow (corresponding to Service B / Resource Server) validates the access token included in the request and performs authorization checks based on configured policies.
```

We leverage this Athenz-based internal platform to implement service-to-service authentication and authorization for MLflow as well.

## Obtaining an Access Token from the Authorization Server
In the Client Credentials Grant Flow, the Authorization Server can authenticate Service A not only by using a client_id and client_secret, but also through mTLS-based authentication.

Athenz provides a feature called [Athenz Copper Argos](https://athenz.github.io/athenz/copper_argos/), which issues client certificates in X.509 format based on the SPIFFE (Secure Production Identity Framework For Everyone) specification—an open standard for secure identity management in distributed systems. These certificates serve as proof of a service’s identity.

By using an Instance Certificate issued through Copper Argos, a service can perform mTLS communication with the Authorization Server and prove that it is a legitimate service.

> In our model training platform, we leverage Athenz Copper Argos so that each Pod performs mTLS authentication with the Authorization Server at startup and automatically obtains an access token. As a result, model training programs running inside the Pod can use the issued access token transparently, without having to implement any token acquisition logic themselves.

Reviewer: ```suggestion
In our model training platform, each Pod uses Athenz Copper Argos to perform mTLS authentication with the Authorization Server at startup and automatically obtains an access token. As a result, model training programs inside the Pod can use the issued access token transparently, without having to implement any token acquisition logic themselves.
```

## Authorization Checks Based on Access Tokens Using the Authorization Proxy
On the MLflow side, we need a mechanism to validate access tokens and perform authorization checks.

> To achieve this, we adopted the [Authorization Proxy](https://github.com/AthenZ/authorization-proxy), one of the products provided by Athenz. The Authorization Proxy runs as a reverse proxy in a Kubernetes sidecar container and is capable of validating access tokens as well as enforcing authorization policies. From the perspective of the OAuth 2.0 specification, this means that the Authorization Proxy takes on the role of performing authentication and authorization checks on behalf of the Resource Server.

Reviewer: ```suggestion
To achieve this, we adopted [Authorization Proxy](https://github.com/AthenZ/authorization-proxy), one of the components provided by Athenz. Authorization Proxy runs as a reverse proxy in a Kubernetes sidecar container, validates access tokens and enforces authorization policies. From an OAuth 2.0 perspective, Authorization Proxy takes the role of performing authentication and authorization checks on behalf of the Resource Server.
```

Because we regularly upgrade MLflow, our policy is to avoid adding custom features directly to the MLflow OSS codebase, as doing so would increase operational complexity. The Authorization Proxy aligns well with this policy, since it allows us to add authentication and authorization features externally without modifying MLflow itself.

> In our deployment, the Authorization Proxy runs as a sidecar attached to each MLflow Pod, and it enforces authentication and authorization for all MLflow API requests.

Reviewer: ```suggestion
In our deployment, each MLflow Pod includes the Authorization Proxy as a sidecar that enforces authentication and authorization for all MLflow API requests.
```

    --data '{"max_results":1}'
HTTP/2 200
content-length: 321
content-type: application/json
{
  "experiments": [
    {
      "experiment_id": "1857",
      "name": "mlflow-example",
      "artifact_location": "s3://sandbox/1857",
      "lifecycle_stage": "active",
      "last_update_time": 1759381587999,
      "creation_time": 1759381587999
    }
  ],
}
```

Example of Unauthorized Access to an MLflow Server
```
# Example of unauthorized access to an MLflow server
 % curl -i -X GET https://${MLFLOW_SERVER_NOT_HAVING_ACCESS_RIGHTS}/api/2.0/mlflow/experiments/search \
    -H 'Content-Type:application/json;'  \
    -H "Authorization: Bearer $(echo $MLFLOW_ACCESS_TOKEN)" \
    --data '{"max_results":1}'
HTTP/2 401
content-length: 0
```
As shown above, when accessing an MLflow server without the proper permissions, the request correctly returns a 401 Unauthorized response — confirming that access control via Athenz works as expected.

With these mechanisms in place, we have built a managed MLflow environment that integrates seamlessly with our Kubernetes-based training platform while ensuring secure and reliable service-to-service usage.

> # MLflow Usage at LY Corporation

Reviewer: @yukimori I wonder we can show this number first before auth sections to let readers get the scale of the platform.
